File: civicbackend/src/main/java/com/civic/civicbackend/service/NotificationService.java
package com.civic.civicbackend.service;

import com.civic.civicbackend.model.Notification;
import com.civic.civicbackend.repository.NotificationRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class NotificationService {

    private final NotificationRepository repo;

    public NotificationService(NotificationRepository repo) {
        this.repo = repo;
    }

    // Create notification for user
    public void createUserNotification(Long userId, String message, String type, Long complaintId) {
        Notification n = new Notification();
        n.setUserId(userId);
        n.setMessage(message);
        n.setType(type);
        n.setComplaintId(complaintId);
        repo.save(n);
    }

    // Create notification for department
    public void createDepartmentNotification(Long departmentId, String message, String type, Long complaintId) {
        Notification n = new Notification();
        n.setDepartmentId(departmentId);
        n.setMessage(message);
        n.setType(type);
        n.setComplaintId(complaintId);
        repo.save(n);
    }

    // Get user notifications sorted by date
    public List<Notification> getUserNotifications(Long userId) {
        return repo.findByUserIdOrderByCreatedAtDesc(userId);
    }

    // Get department notifications sorted by date
    public List<Notification> getDepartmentNotifications(Long departmentId) {
        return repo.findByDepartmentIdOrderByCreatedAtDesc(departmentId);
    }

    // Get unread count for user
    public long getUnreadUserCount(Long userId) {
        return repo.countByUserIdAndIsReadFalse(userId);
    }

    // Get unread count for department
    public long getUnreadDepartmentCount(Long departmentId) {
        return repo.countByDepartmentIdAndIsReadFalse(departmentId);
    }

    // Mark notification as read
    public void markAsRead(Long notificationId) {
        repo.findById(notificationId).ifPresent(n -> {
            n.setRead(true);
            repo.save(n);
        });
    }

    // Mark all user notifications as read
    public void markAllUserAsRead(Long userId) {
        List<Notification> notifications = repo.findByUserId(userId);
        notifications.forEach(n -> n.setRead(true));
        repo.saveAll(notifications);
    }

    // Mark all department notifications as read
    public void markAllDepartmentAsRead(Long departmentId) {
        List<Notification> notifications = repo.findByDepartmentId(departmentId);
        notifications.forEach(n -> n.setRead(true));
        repo.saveAll(notifications);
    }
}

File: civic_api/app/services/notification_service.py
from sqlalchemy.orm import Session

from app.models import Notification


def create_user_notification(db: Session, user_id: int, message: str, type_: str, complaint_id: int):
    notification = Notification(
        user_id=user_id,
        message=message,
        type=type_,
        complaint_id=complaint_id,
    )
    db.add(notification)
    db.commit()


def create_department_notification(
    db: Session, department_id: int, message: str, type_: str, complaint_id: int
):
    notification = Notification(
        department_id=department_id,
        message=message,
        type=type_,
        complaint_id=complaint_id,
    )
    db.add(notification)
    db.commit()


def get_user_notifications(db: Session, user_id: int) -> list[Notification]:
    return (
        db.query(Notification)
        .filter(Notification.user_id == user_id)
        .order_by(Notification.created_at.desc())
        .all()
    )


def get_department_notifications(db: Session, department_id: int) -> list[Notification]:
    return (
        db.query(Notification)
        .filter(Notification.department_id == department_id)
        .order_by(Notification.created_at.desc())
        .all()
    )


def get_unread_user_count(db: Session, user_id: int) -> int:
    return db.query(Notification).filter(Notification.user_id == user_id, Notification.is_read.is_(False)).count()


def get_unread_department_count(db: Session, department_id: int) -> int:
    return (
        db.query(Notification)
        .filter(Notification.department_id == department_id, Notification.is_read.is_(False))
        .count()
    )


def mark_as_read(db: Session, notification_id: int):
    notification = db.query(Notification).filter(Notification.id == notification_id).first()
    if notification:
        notification.is_read = True
        db.commit()


def mark_all_user_as_read(db: Session, user_id: int):
    notifications = db.query(Notification).filter(Notification.user_id == user_id).all()
    for notification in notifications:
        notification.is_read = True
    db.commit()


def mark_all_department_as_read(db: Session, department_id: int):
    notifications = db.query(Notification).filter(Notification.department_id == department_id).all()
    for notification in notifications:
        notification.is_read = True
    db.commit()
